src/main/java/io/confluent/connect/jdbc/sink/JdbcSinkTask.java

```diff
@@ -86,19 +101,39 @@ public void put(Collection<SinkRecord> records) {
       }
       SQLException sqlAllMessagesException = new SQLException(sqleAllMessages);
       sqlAllMessagesException.setNextException(sqle);
-      if (remainingRetries == 0) {
-        throw new ConnectException(sqlAllMessagesException);
-      } else {
+      if (remainingRetries > 0) {
         writer.closeQuietly();
         initWriter();
         remainingRetries--;
         context.timeout(config.retryBackoffMs);
         throw new RetriableException(sqlAllMessagesException);
+      } else {
+        if (reporter != null) {
+          unrollAndRetry(records);
+        } else {
+          throw new ConnectException(sqlAllMessagesException);
+        }
       }
     }
     remainingRetries = config.maxRetries;
   }
 
+  private void unrollAndRetry(Collection<SinkRecord> records) {
+    for (SinkRecord record : records) {
+      try {
+        writer.write(Collections.singletonList(record));
+      } catch (SQLException sqle) {
+        String sqleAllMessages = "Exception chain:" + System.lineSeparator();
```

Comment:
Due to a new PR merged which addresses some logging, it is messy to make this code chunk into a separate function since we want two pieces of information from it now.

Reply:
You can still refactor the code with the side-effect (debug call), you're calling it on a different error here. It's possible that each individual row could return multiple errors, and it would make sense to debug log that.

Reply:
https://github.com/confluentinc/kafka-connect-jdbc/blob/10.0.x/src/main/java/io/confluent/connect/jdbc/sink/JdbcDbWriter.java#L57

Autocommit for transaction is set to false. Basically when you run into some sql exception for e.g PK constraints, connection will be with invalid transaction, and all subsequent record will throw exception and end up in dlq.

I think if we can put there transaction abort/rollback or standard writer.closeQuietly(); method would address this issue.

Reply:
I agree with @gharris1727 , we should try to reduce code duplication here. Even with the latest changes in `master`, the only addition is a `totalExceptions` , which can be part of the returned value on the helper method. Or, just one-off count total exceptions again outside of the helper method.

Reply:
@ncliang I refactored the method, let me know what you think